Single file workspaces

natefaubion opened the pull request:
Fixes #1275

cryogenian commented on src/SlamData/Workspace/Deck/Component/State.purs:
```diff
@@ -82,11 +78,8 @@ import Quasar.Advanced.Types (ProviderR)
 import SlamData.Workspace.Deck.Gripper.Def (GripperDef)
-import SlamData.Workspace.StateMode (StateMode(..))
 
 import Utils (censor)
```
I know that this had been introduced previously, but could you rename this function? `censor` is a bit confusing, like where is `Writer` here? :)

natefaubion replied:
Yeah, I think this is called `hush` in Haskell-land.

cryogenian commented on src/SlamData/Workspace/Eval/Persistence.purs:
```diff
@@ -509,224 +431,122 @@ groupDeck orn bias deckId newParentId newParentCoord = runExceptT do
+    , state: Nothing
+    , sources: Set.empty
+    , tick: Nothing
+    } ∷ Card.Cell
```
Is this annotation necessary because of `Nothing`s?

natefaubion replied:
It's because of inference around records and polymorphic members. It has trouble with the `Set t1` introduced by `Set.empty`.

cryogenian replied:
Got it, thanks

cryogenian commented on src/SlamData/Workspace/Legacy.purs:
```diff
@@ -0,0 +1,198 @@
+    )
+  ⇒ DirPath
+  → m (Either QE.QError Current.Workspace)
+loadCompatWorkspace path = runExceptT do
```
Maybe this should immediately save new deck json after loading and remove all redundant files?

natefaubion replied:
I thought about that, and I can do that in the case of proper edit permissions. It's possible that someone can be viewing it without the permissions to update it, in which case that doesn't work.

cryogenian replied:
Maybe just try to update it then?

natefaubion replied:
How is the `.tmp` directory laid out? Could I just blast away the whole `.tmp` directory, and people still keep their permissions for sharing?

cryogenian replied:
.tmp -> deckId -> tmp file 
Most of tmp files are view mounts, some (cache card outputs) are files. If you remove .tmp permissions user can't see anything in deck I think

cryogenian commented on src/SlamData/Workspace/Deck/Component.purs:
```diff
@@ -463,127 +424,97 @@ presentReason input cardType =
   st ← H.get
-  res ← liftH' $ P.deleteDeck st.id
-  for_ res (maybe navigateToIndex navigateToDeck)
+  _ ← liftH' $ P.deleteDeck opts.deckId
```
`_`

cryogenian commented on src/SlamData/Workspace/Deck/Component.purs:
```diff
@@ -463,127 +424,97 @@ presentReason input cardType =
   activeCardIndex st = do
     { cache } ← liftH' Wiring.expose
-    map _.cardIndex <$> H.liftH (H.liftH (Cache.get st.id cache.activeState))
+    map _.cardIndex <$> H.liftH (H.liftH (Cache.get opts.deckId cache.activeState))
```
`liftH'`

cryogenian commented on src/SlamData/Workspace/Deck/Component/State.purs:
```diff
@@ -122,17 +115,15 @@ derive instance eqResponsiveSize ∷ Eq ResponsiveSize
-  , parent ∷ Maybe (DeckId × CardId)
-  , stateMode ∷ StateMode
+  { name ∷ String
+  , stateMode ∷ Maybe String
```
Why this should be `Maybe String`?

natefaubion replied:
I guess this should be `loadError` or something. The full `StateMode` type is unnecessary.

cryogenian replied:
This means that `StateMode` is redundant now, right?

natefaubion replied:
It's used in `Workspace.Component` instead for loading the workspace model, but I think we can remove a constructor (`Preparing` makes no sense).

cryogenian commented on src/SlamData/Workspace/Eval/Traverse.purs:
```diff
@@ -16,108 +16,174 @@ limitations under the License.
+        TraverseDeck
+          { deckId
+          , deck
+          , cards: List.catMaybes (unfoldCard <$> List.fromFoldable (getDeck deck).cards)
```
Why not?
```purescript
List.mapMaybe unfoldCard $ List.fromFoldable (getDeck deck).cards
```
or something like this 😛 
```
foldMap (List.fromFoldable <<< unfoldCard) $ (getDeck deck).cards
```
That's not really an issue, just wondering.

natefaubion replied:
Probably because I rewrote it a few times :P I'll change it to `mapMaybe`.

cryogenian commented:
That's really cool and even easier to understand then `eval-rewrite` 💯  Just couple minor things, like redundant `_`

natefaubion commented on src/SlamData/Workspace/Component.purs:
```diff
@@ -249,9 +302,9 @@ runFreshWorkspace cards = do
-  setRoot deckId
-  urlVarMaps ← liftH' $ Cache.snapshot varMaps
-  navigate $ WorkspaceRoute path (Just deckId) (WA.Load accessType) urlVarMaps
+  _ ← liftH' P.saveWorkspace
```
I am using `_ <- foo` because the action returns a useful result, but in this case I don't care about it.

natefaubion replied:
I personally don't care either way, but I know there's "some consensus" in the community that you should use `_` if the return type isn't `Unit`.

cryogenian replied:
I've seen this before actually and don't have any strong feelings about it. It just seems that we don't use this approach anywhere because any `H.query blablabla` returns `Maybe`

natefaubion replied:
👍

natefaubion commented:
* Changed `stateMode` to `loadError` in Deck
* Removed the `Preparing` constructor from `StateMode` (we still use StateMode in Workspace)
* Prunes legacy data from the FS
* Added comments about cursors
* Style changes
